### PR TITLE
Issue 16277

### PR DESCRIPTION
**Pull request by Nekmo**
- [X] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [X] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

- [X] Bug fix

---

Fixed Issue #16277: Atresplayer broken: ERROR: Unsupported URL

### CODE REVIEW

**Comment by Nekmo** [comment]
Fixed Python2 support :( (urllib.error)

**Comment by Nekmo** [comment]
Please check again "do-not-merge" label cc/ @dstftw

**Comment by Aleixenandros** [comment]
¿Por qué no lo implementan en las nuevas versiones?

Funciona perfecto desde el repositorio de @Nekmo

File: youtube_dl/extractor/atresplayer.py [hunk 1]
```diff
@@ -1,74 +1,59 @@
...
-import time
-import hmac
-import hashlib
-import re
+import json
+
+try:
+    from urllib.error import HTTPError
+except ImportError:
+    from urllib2 import HTTPError
```
Comment (dstftw) [inline youtube_dl/extractor/atresplayer.py]: `compat_HTTPError`.

File: youtube_dl/extractor/atresplayer.py [hunk 2]
```diff
@@ -1,74 +1,59 @@
...
+try:
+    from json import JSONDecodeError
+except ImportError:
+    JSONDecodeError = ValueError
+
 
 class AtresPlayerIE(InfoExtractor):
-    _VALID_URL = r'https?://(?:www\.)?atresplayer\.com/television/[^/]+/[^/]+/[^/]+/(?P<id>.+?)_\d+\.html'
+    _VALID_URL = r'https?://(?:www\.)?atresplayer\.com/[^/]+/[^/]+/' \
+                 r'[^/]+/[^/]+/[^/_]+_(?P<id>[A-z0-9]+)/?'
```
Comment (dstftw) [inline youtube_dl/extractor/atresplayer.py]: `/?` is pointless at the end.

Comment (dstftw) [inline youtube_dl/extractor/atresplayer.py]: Keep the regex on a single line.

Comment (dstftw) [inline youtube_dl/extractor/atresplayer.py]: `A-z` incorrect.

File: youtube_dl/extractor/atresplayer.py [hunk 3]
```diff
@@ -1,74 +1,59 @@
...
     _NETRC_MACHINE = 'atresplayer'
     _TESTS = [
         {
-            'url': 'http://www.atresplayer.com/television/programas/el-club-de-la-comedia/temporada-4/capitulo-10-especial-solidario-nochebuena_2014122100174.html',
-            'md5': 'efd56753cda1bb64df52a3074f62e38a',
+            'url': 'https://www.atresplayer.com/lasexta/programas/el-'
```
Comment (dstftw) [inline youtube_dl/extractor/atresplayer.py]: Don't split. Same everywhere.

File: youtube_dl/extractor/atresplayer.py [hunk 4]
```diff
@@ -79,124 +64,83 @@ def _login(self):
...
 
         request = sanitized_Request(
-            self._URL_VIDEO_TEMPLATE.format('windows', episode_id, timestamp_shifted, token),
+            video_url,
             headers={'User-Agent': self._USER_AGENT})
+        try:
+            video_data = self._download_json(request, video_id,
+                                             'Downloading video JSON',
+                                             fatal=True)
+        except ExtractorError as e:
```
Comment (dstftw) [inline youtube_dl/extractor/atresplayer.py]: `expected_status` to `_download_json` instead.

File: youtube_dl/extractor/atresplayer.py [hunk 5]
```diff
@@ -79,124 +64,83 @@ def _login(self):
...
-        error = self._html_search_regex(
-            r'(?s)<ul[^>]+class="[^"]*\blist_error\b[^"]*">(.+?)</ul>',
-            response, 'error', default=None)
-        if error:
-            raise ExtractorError(
-                'Unable to login: %s' % error, expected=True)
+        try:
+            response = self._download_json(
+                request, None, 'post to login form')
+        except ExtractorError as e:
```
Comment (dstftw) [inline youtube_dl/extractor/atresplayer.py]: `expected_status` to `_download_json` instead.

File: youtube_dl/extractor/atresplayer.py [hunk 6]
```diff
@@ -79,124 +64,83 @@ def _login(self):
...
 
         login_form = {
-            'j_username': username,
-            'j_password': password,
+            'username': username,
+            'password': password,
         }
 
+        self._download_webpage(self._LOGIN_URL, None, 'get login page')
         request = sanitized_Request(
```
Comment (dstftw) [inline youtube_dl/extractor/atresplayer.py]: Inline into actual `_download_*` call.

File: youtube_dl/extractor/atresplayer.py [hunk 7]
```diff
@@ -79,124 +64,83 @@ def _login(self):
...
+            self._download_webpage(response['targetUrl'], None,
+                                   'Set login session')
+
+    def _atres_player_error(self, body_response, original_exception):
+        try:
+            data = json.loads(body_response)
+        except JSONDecodeError:
+            return original_exception
+        if isinstance(data, dict) and 'error' in data:
+            return ExtractorError('{} returned error: {} ({})'.format(
```
Comment (dstftw) [inline youtube_dl/extractor/atresplayer.py]: No `{}`.

File: youtube_dl/extractor/atresplayer.py [hunk 8]
```diff
@@ -79,124 +64,83 @@ def _login(self):
...
+            video_data = self._download_json(request, video_id,
+                                             'Downloading video JSON',
+                                             fatal=True)
+        except ExtractorError as e:
+            if len(e.exc_info) <= 1 or e.exc_info[1].code != 403:
+                raise
+            raise self._atres_player_error(e.exc_info[1].file.read(), e)
+
+        for source in video_data['sources']:
+            if source['type'] == "application/dash+xml":
```
Comment (dstftw) [inline youtube_dl/extractor/atresplayer.py]: Single quotes.

Comment (dstftw) [inline youtube_dl/extractor/atresplayer.py]: Should not break if no `type`.

File: youtube_dl/extractor/atresplayer.py [hunk 9]
```diff
@@ -79,124 +64,83 @@ def _login(self):
...
-            }]
-
         return {
             'id': video_id,
-            'title': title,
-            'description': description,
-            'thumbnail': thumbnail,
-            'duration': duration,
+            'title': video_data['titulo'],
+            'description': video_data['descripcion'],
```
Comment (dstftw) [inline youtube_dl/extractor/atresplayer.py]: Read coding conventions on optional/mandatory meta fields.

File: youtube_dl/extractor/atresplayer.py (hunk 6 above)
Comment (Nekmo) [reply youtube_dl/extractor/atresplayer.py]: I don0t understand this comment.

File: youtube_dl/extractor/atresplayer.py (hunk 4 above)
Comment (Nekmo) [reply youtube_dl/extractor/atresplayer.py]: I don't find "expected_status".

File: youtube_dl/extractor/atresplayer.py (hunk 6 above)
Comment (bato3) [reply youtube_dl/extractor/atresplayer.py]: @Nekmo You can make json POST with `_download_json`, and set expected statuses (`data=urlencode_postdata(form_data)`)

From common.py
``` python
      def _download_json(
            self, url_or_request, video_id, note='Downloading JSON metadata',
            errnote='Unable to download JSON metadata', transform_source=None,
            fatal=True, encoding=None, data=None, headers={}, query={},
            expected_status=None):
        """
        Return the JSON object as a dict.

        See _download_webpage docstring for arguments specification.
        """
```

File: youtube_dl/extractor/atresplayer.py (hunk 8 above)
Comment (Nekmo) [reply youtube_dl/extractor/atresplayer.py]: Already fixed on PR bbb857cc948d7ec88398d4b31383df77b69b3a53

File: youtube_dl/extractor/atresplayer.py (hunk 6 above)
Comment (Nekmo) [reply youtube_dl/extractor/atresplayer.py]: The response from the server is not a json. This request is to set cookies.

**Comment by pereorga** [comment]
FWIW, this is working fine locally, on both Windows 10 (Python 3.7.3) and Debian stable (Python 3.5.3).


It does not work, however, on an Ubuntu 16.04 instance that I have in DigitalOcean (Python 3.5.2). I guess it may be because of IP address-based geo-blocking? I am getting an HTTP 403:

```
~/dev/youtube-dl/youtube_dl$ python3 ./__main__.py -uUSERNAME -pPASSWORD https://www.atresplayer.com/lasexta/programas/salvados/temporada-14/francisco_5c9f49237ed1a885b9056c1f/

[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['--verbose', '-uUSERNAME', '-pPASSWORD', 'https://www.atresplayer.com/lasexta/programas/salvados/temporada-14/francisco_5c9f49237ed1a885b9056c1f/']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2018.06.11
[debug] Git HEAD: bbb857c
[debug] Python version 3.5.2 (CPython) - Linux-4.4.0-143-generic-x86_64-with-Ubuntu-16.04-xenial
[debug] exe versions: none
[debug] Proxy map: {}
[AtresPlayer] get login page
[AtresPlayer] post to login form
[AtresPlayer] Set login session
[AtresPlayer] 5c9f49237ed1a885b9056c1f: Downloading player JSON
[AtresPlayer] 5c9f49237ed1a885b9056c1f: Downloading video JSON
Traceback (most recent call last):
  File "/home/netol/dev/youtube-dl/youtube_dl/extractor/common.py", line 579, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "/home/netol/dev/youtube-dl/youtube_dl/YoutubeDL.py", line 2211, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/lib/python3.5/urllib/request.py", line 472, in open
    response = meth(req, response)
  File "/usr/lib/python3.5/urllib/request.py", line 582, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python3.5/urllib/request.py", line 510, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.5/urllib/request.py", line 444, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.5/urllib/request.py", line 590, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 403: Forbidden

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/netol/dev/youtube-dl/youtube_dl/extractor/atresplayer.py", line 109, in _real_extract
    fatal=True)
  File "/home/netol/dev/youtube-dl/youtube_dl/extractor/common.py", line 767, in _download_json
    data=data, headers=headers, query=query)
  File "/home/netol/dev/youtube-dl/youtube_dl/extractor/common.py", line 752, in _download_json_handle
    encoding=encoding, data=data, headers=headers, query=query)
  File "/home/netol/dev/youtube-dl/youtube_dl/extractor/common.py", line 599, in _download_webpage_handle
    urlh = self._request_webpage(url_or_request, video_id, note, errnote, fatal, data=data, headers=headers, query=query)
  File "/home/netol/dev/youtube-dl/youtube_dl/extractor/common.py", line 588, in _request_webpage
    raise ExtractorError(errmsg, sys.exc_info()[2], cause=err)
youtube_dl.utils.ExtractorError: Unable to download JSON metadata: HTTP Error 403: Forbidden (caused by <HTTPError 403: 'Forbidden'>); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./__main__.py", line 19, in <module>
    youtube_dl.main()
  File "/home/netol/dev/youtube-dl/youtube_dl/__init__.py", line 472, in main
    _real_main(argv)
  File "/home/netol/dev/youtube-dl/youtube_dl/__init__.py", line 462, in _real_main
    retcode = ydl.download(all_urls)
  File "/home/netol/dev/youtube-dl/youtube_dl/YoutubeDL.py", line 2001, in download
    url, force_generic_extractor=self.params.get('force_generic_extractor', False))
  File "/home/netol/dev/youtube-dl/youtube_dl/YoutubeDL.py", line 792, in extract_info
    ie_result = ie.extract(url)
  File "/home/netol/dev/youtube-dl/youtube_dl/extractor/common.py", line 500, in extract
    ie_result = self._real_extract(url)
  File "/home/netol/dev/youtube-dl/youtube_dl/extractor/atresplayer.py", line 113, in _real_extract
    raise self._atres_player_error(e.exc_info[1].file.read(), e)
  File "/home/netol/dev/youtube-dl/youtube_dl/extractor/atresplayer.py", line 80, in _atres_player_error
    data = json.loads(body_response)
  File "/usr/lib/python3.5/json/__init__.py", line 312, in loads
    s.__class__.__name__))
TypeError: the JSON object must be str, not 'bytes'
```

**Pull request closed without merging**